### PR TITLE
v1_4_getESDTLocalRoles

### PR DESCRIPTION
**Pull request by alyn509**

### CODE REVIEW

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -914,6 +915,26 @@ func v1_4_getESDTTokenData(
 	return int32(len(esdtData.Value.Bytes()))
 }
 
+func v1_4_getESDTLocalRoles(context unsafe.Pointer, tokenIDOffset int32, tokenIDLen int32) int32 {
```
Comment (andrei-marinica) [inline arwen/elrondapi/elrondei.go:918]: Pass token id as managed buffer, so only a handle here.
Also, return i64 (long long).

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -914,6 +915,26 @@ func v1_4_getESDTTokenData(
...
+	runtime := arwen.GetRuntimeContext(context)
+	storage := arwen.GetStorageContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.StorageLoad
+	metering.UseGasAndAddTracedGas(storageLoadLengthName, gasToUse)
+
+	tokenID, err := runtime.MemLoad(tokenIDOffset, tokenIDLen)
+	if err != nil {
+		return 0
```
Comment (andrei-diaconescu) [inline arwen/elrondapi/elrondei.go:928]: arwen.WithFault()

**Review by andrei-diaconescu — DISMISSED** [review]

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -914,6 +920,28 @@ func v1_4_getESDTTokenData(
...
+	runtime := arwen.GetRuntimeContext(context)
+	storage := arwen.GetStorageContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.StorageLoad
+	metering.UseGasAndAddTracedGas(storageLoadLengthName, gasToUse)
+
+	tokenID, err := managedType.GetBytes(tokenIdHandle)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return 0
```
Comment (andrei-diaconescu) [inline arwen/elrondapi/elrondei.go:935]: can the final `return int64(len(data))` be 0? If so, then change the return value on this `if arwen.WithFault()`

Comment (alyn509) [reply arwen/elrondapi/elrondei.go:935]: 0 is considered NONE, and the reasoning is that if it is not valid it doesn't have roles

Comment (andrei-diaconescu) [reply arwen/elrondapi/elrondei.go:935]: better return -1 on `if arwen.WithFault`, not 0

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -914,6 +932,84 @@ func v1_4_getESDTTokenData(
...
+	}
+	tokenID, err := runtime.MemLoad(tokenIDOffset, tokenIDLen)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return -1
+	}
+
+	// ticker must be all uppercase alphanumeric
+	tickerLen := tokenIDLen - additionalRandomCharsLength
+	for i := int32(0); i < tickerLen; i++ {
```
Comment (andrei-marinica) [inline arwen/elrondapi/elrondei.go:993]: Please extract the validation logic to a separate function.

File: arwen/elrondapi/elrondei.go [hunk 5]
```diff
@@ -50,6 +50,8 @@ package elrondapi
 // extern int32_t	v1_4_getESDTNFTAttributeLength(void *context, int32_t addressOffset, int32_t tokenIDOffset, int32_t tokenIDLen, long long nonce);
 // extern int32_t	v1_4_getESDTNFTURILength(void *context, int32_t addressOffset, int32_t tokenIDOffset, int32_t tokenIDLen, long long nonce);
 // extern int32_t	v1_4_getESDTTokenData(void *context, int32_t addressOffset, int32_t tokenIDOffset, int32_t tokenIDLen, long long nonce, int32_t valueOffset, int32_t propertiesOffset, int32_t hashOffset, int32_t nameOffset, int32_t attributesOffset, int32_t creatorOffset, int32_t royaltiesOffset, int32_t urisOffset);
+// extern long long	v1_4_getESDTLocalRoles(void *context, int32_t tokenIdHandle);
+// extern int32_t	v1_4_validateTokenIdentifier(void *context, int32_t tokenIDOffset, int32_t tokenIDLen);
```
Comment (andrei-marinica) [inline arwen/elrondapi/elrondei.go:54]: Use a managed buffer as input.
v1_4_validateTokenIdentifier(void *context, int32_t int32_t tokenIdHandle);

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Pull request closed without merging**